doc/developer/guide-changes.md

```diff
@@ -0,0 +1,699 @@
+
+The first step to submitting a change to Materialize is, of course, to actually
+write the code involved. See earlier sections of the developer guide for help
+with the logistics, like cloning the repository and running the test suites.
+
+When you are ready to submit your change for review, the first step is to break
+it up into commits. There are two schools of thought regarding commit hygiene.
+
+  * **Semantic commits.** In the semantic commits approach, *each commit* is the
+    unit of review. Every commit should have a clear purpose and a well-written
+    commit message.
+
+    If you use this approach, you simply write "see commit messages for details"
+    in your PR description, though giving a quick overview can be helpful if
+    there are many commits involved. When you merge a PR, use the standard merge
+    strategy so that the semantic commits are preserved on the `main` branch.
+
+    For an example of this approach, see PR [#3982].
+
+  * **Semantic PRs.** In the semantic PRs approach, your *entire PR* is the unit
+    of review, and there is no meaning to the division of commits. It is common
+    to simply commit as you go with this approach, with messages like "feature
+    works", "tests passing", "apply cargo fmt and clippy."
+
+    If you use the semantic PRs approach, apply your explanatory efforts to
+    writing a good PR description, rather than writing good commit messages.
+    Then, be sure to use the "squash" strategy when you merge. That way, your
+    commits will be replaced with a neat, squashed commit on the `main` branch
+    containing your PR description when you merge.
+
+    For an example of this approach, see PR [#3808].
+
+You should endeavor to limit each *review unit* (RU) to just one semantic
+change. When a single RU contains multiple unrelated changes, it becomes much
+harder for your reviewer to understand, and the quality of the review suffers.
+
+A good test of whether you've included just "one thing" in your change is to
+try to write the description for the change, as described in the following
+section. If your description reads like "do X and do Y and do Z", you've tried
+to do three things, rather than just one.
+
+A good example of what *not* to do is fixing a bug and adding a feature in the
+same RU. This happens frequently, as often the process of adding a new feature
+requires fixing a bug, and it is quite convenient to fix that bug in the same
+RU. But doing so makes it very difficult to view the code for the bugfix or
+feature in isolation. This can be particularly problematic if the bugfix or
+feature needs to be reverted—it is much more difficult to revert just one or the
+other if they are in the same RU.
+
+### Change descriptions
+
+Writing good commit messages (or PR descriptions, if you use the semantic PRs
+strategy) is essential to leaving a public record of *why* a change was made to
+the codebase.
+
+Most Git commit conventions agree on at least the following form:
```

Comment:
this is a good place to throw in some sample commits that are good examples and maybe also link to the wiki page for imperative tense?

Reply:
Done.